File: Arrays/2sum.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        int n = nums.size();
        for (int i = 0; i < n; i++) {
            for (int j = i+1; j < n; j++) {
                if (nums[i] + nums[j] == target) {
                    return {i, j};
                }
            }
        }
        return {-1, -1};
    }
};

class SolutionOptimized {
public: 
     vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int,int>mp;
        for(int i=0;i<nums.size();i++){
            int complement=target-nums[i]; // Find the complement that adds up to the target
            if(mp.find(complement)!=mp.end()){ // find searches for key in map and returns iterator to end if not found
                return {mp[complement],i}; // Return the indices of the complement and the current number
            }
            mp[nums[i]]=i; // we store index as value and number as key
        }
        return {-1,-1};
     }
};

int main(){
    Solution sol;
    vector<int> nums = {2, 7, 11, 15};
    int target = 9;
    vector<int> result = sol.twoSum(nums, target);
    for(int index : result){
        cout << index << " ";
    }

    SolutionOptimized solOpt;
    vector<int> resultOpt = solOpt.twoSum(nums, target);
    cout << endl;
    for(int index : resultOpt){
        cout << index << " ";
    }
    return 0;
}

File: Arrays/2_sum.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        pair_indices = {} # we will use a dictionary to store the indices of the numbers we have seen so far
        for i, num in enumerate(nums): # enumerate will give us both the index and the number at that index
            complement = target - num
            if complement in pair_indices: # if the complement is in the dictionary it means we have found a pair of numbers that add up to the target
                return [pair_indices[complement], i] # we return the indices of the complement and the current number
            pair_indices[num] = i # else we add the current number and its index to the dictionary
        return [] # if we have gone through the entire list and haven't found a pair we return
    
# Example usage
sol = Solution()
nums = [2, 7, 11, 15]
target = 9
print(sol.twoSum(nums, target)) # Output: [0, 1] because nums[0] + nums[1] = 2 + 7 = 9